File: Dota2 Senate.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        queue<int> d,r;
        int n=senate.length();
        for (int i=0;i<n;i++)
        {
            if (senate[i]=='R')
            r.push(i);
            else
            d.push(i);
        }
        while(!d.empty() && !r.empty())
        {
          if (d.front()>r.front())
            r.push(n++);
            else
            d.push(n++);  
        r.pop(),d.pop();
         }
         return (r.empty())?("Dire"):("Radiant");
    }
};

File: camp/week-2/leetcode/dota2-senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        freq = defaultdict(int)
        senate = list(senate)
        R = []
        D = []
        for i in senate:
            freq[i] += 1
        j = 0
        while freq["R"] > 0 and freq["D"] > 0:
            if senate[j] == "R":
                if len(R) > 0:
                    senate[j] = "N"
                    R.pop()
                else:
                    freq["D"] -=1
                    D.append("N")
            elif senate[j] == "D":
                if len(D) > 0:
                    senate[j] = "N"
                    D.pop()
                else:
                    freq["R"] -=1
                    R.append("N") 
            j+=1
            if j == len(senate):
                j = 0
        return "Radiant" if freq["D"] <= 0 else "Dire"